### ISSUE TITLE
The Braid Handler will now ignore non-perennial streams, if a perennial stream network was given to the BRAT Table tool

### ISSUE THREAD

**Issue by banderson1618** [Riverscapes/pyBRAT#260]
I've successfully implemented what we talked about on Monday. The braid handler now doesn't include non-perennial streams when looking for clusters.

This also means that the BRAT Table now creates a field called `IsPeren` that is assigned a 1 if the stream shares a line segment with the perennial network, and a 0 if it doesn't.

**Comment by bangen** [comment]
@banderson1618 Here's [a link](https://usu.box.com/s/l2epj4q9bf7jeiinhkfbpzx2q4kkhhz0) to some data to test the code on.  Truckee is from TNC project and Lemhi is from Idaho project.

**Comment by banderson1618** [comment]
Thanks, @bangen, I'll run those and let you know if I run into any problems.

**Comment by bangen** [comment]
@banderson1618 @wally-mac 

I was just running the Truckee through the BRAT Table tool in order to test our proposed changes to the iHyd script.  There are a ton of segments that have a 'IsMultiCh' value of 1 that aren't part of the perennial network (see screenshots below).  @banderson1618, based on looking at the attribute table it looks like you're assigning non-perennial segments a 'ClusterID' of -1 (which is good).  You should also update the 'IsMultiCh' field and set it to 0.  As it stands it's a little misleading and if people edit based on that field it'll be a bunch of wasted time.   

I'm going to re-open the issue. Once you fix it and push it to the main branch go ahead and re-close.

**Perennial:** 
![Truckee_Perennial](https://user-images.githubusercontent.com/6502954/55199026-ad2de380-5175-11e9-8151-1ec92aa51729.PNG)

**IsMultCh:** 
![Truckee_MultiCh_Issue](https://user-images.githubusercontent.com/6502954/55199037-bcad2c80-5175-11e9-924e-6142471050d3.PNG)

**Comment by bangen** [comment]
@banderson1618 @wally-mac 

Ok.  I took a look and there's definitely a larger issue here.  If you look at the screenshots below the pink lines have ClusterID > 0.  @banderson1618 looking at the output it looks like you're clipping the clusters to the perennial network rather than running the cluster code on just the perennial network.  Please fix asap.

![Truckee_MultiCh_Issue_ClusterID](https://user-images.githubusercontent.com/6502954/55199256-afdd0880-5176-11e9-99a4-3cdafafb8e19.PNG)

**Zoomed View:**
![Truckee_MultiCh_Issue_ClusterID_Zommed](https://user-images.githubusercontent.com/6502954/55199273-c1261500-5176-11e9-9616-b63b4c9fa606.PNG)

**Comment by banderson1618** [comment]
The primary goal for the perennial check was to be a quick fix that prevents heavily braided sections of river from forming into mega-clusters. A more comprehensive fix would involve significant changes to the way we currently calculate isMultiCh, similar to what we did with excluding canals from the isMultiCh calculation. Because I have limited time left in the office, and we have a limited amount of time before we need to batch process, I decided to go with a more simple fix. If @bangen and @wally-mac think this is high enough priority, I can invest time in a solution, but I only have a month before I leave, and this might consume a week of that time, if it takes about the same amount of time as the canals fix took. From my perspective, this doesn't seem that critical, but I'll defer to your judgement.

**Comment by bangen** [comment]
@banderson1618 Can you provide @wally-mac and I with an update on where you're at with the changes to this?

**Comment by banderson1618** [comment]
Yeah, thanks for the reminder. I did the changes, and they seem to be working, but I didn't get a chance to test them out extensively. I'll try to do that next week and send you what results from that.

**Comment by bangen** [comment]
Can you please send links to the data that you did test and let me know which branch the changes are on?   We're trying to batch process so need the fix in place sooner than later.   If the changes look good on my end and you can't get to pushing until next week then I can take care of it.  Just let me know.

**Comment by banderson1618** [comment]
Here you go!

[BRAT_Table.zip](https://github.com/Riverscapes/pyBRAT/files/3074886/BRAT_Table.zip)

**Comment by bangen** [comment]
@banderson1618  Thanks.  What branch are the changes on?

**Comment by bangen** [comment]
also, what basin did you send?

**Comment by banderson1618** [comment]
I'm not sure what basin, but the changes are just on the master branch
(since I had already merged the original braid changes into master).

**Comment by bangen** [comment]
@banderson1618 this is looking a lot better.  There's still something that seems a little off (see screenshots below).  Any idea why the circled area isn't being correctly classified?  I noticed a few other spots like this in the basin.

![IsPerennial](https://user-images.githubusercontent.com/6502954/56062279-b5af1e00-5d20-11e9-8820-7258ae3d1cc2.PNG)
![IsMutliCh](https://user-images.githubusercontent.com/6502954/56062285-b8117800-5d20-11e9-9dbf-8ea67f70edaf.PNG)

**Comment by banderson1618** [comment]
@bangen, I ran it again, but with the Donut Polygons saved to disk. It looks like there's a problem with the Select Layer By Location, since the Donut Polygon was created properly. I'm confused about how this problem appears here, but not otherwise. Maybe it's a problem with the perennial network, that the polygon is based off of, is slightly different from the base layer? This might be an input problem, I'll look into it more.

![Capture](https://user-images.githubusercontent.com/21353496/56155794-f6a16f80-5f78-11e9-931b-38b9f1914926.PNG)

**Comment by banderson1618** [comment]
As best I can tell, this weirdness is due to some odd interaction with the handleCanals function, since it seems to go away when I don't give the program a canal input. I'll continue looking for a solution.

**Comment by banderson1618** [comment]
[BRAT_Table.zip](https://github.com/Riverscapes/pyBRAT/files/3091729/BRAT_Table.zip)
I think I fixed the canal problem, this should look better.

**Comment by bangen** [comment]
@banderson1618 

Please take a look at your code below.  You're setting both the `stream_network_no_canals` and `perennial_no_canals` to the exact same path (...so overwriting the former if input perennial network exists).  Is this what you intended?

```
def handleCanals(stream_network, canal, perennial_network, temp_folder, is_verbose):
    """
    Finds braided sections of the stream network, not counting canals, if canals are available
    :param stream_network:
    :param canal:
    :param temp_folder:
    :param is_verbose:
    :return:
    """
    if is_verbose:
        arcpy.AddMessage("Removing canals...")
    if arcpy.GetInstallInfo()['Version'][0:4] == '10.5':
        stream_network_no_canals = os.path.join(temp_folder, "NoCanals.shp")
        perennial_no_canals = os.path.join(temp_folder, "NoCanals.shp")
    else:
        stream_network_no_canals = os.path.join('in_memory', 'NoCanals')
        perennial_no_canals = os.path.join(temp_folder, "NoCanals.shp")

    arcpy.Erase_analysis(stream_network, canal, stream_network_no_canals)
    if perennial_network is not None:
        arcpy.Erase_analysis(perennial_network, canal, perennial_no_canals)
    findBraidedReaches(stream_network_no_canals, perennial_no_canals, is_verbose)
```

**Comment by banderson1618** [comment]
Thanks for pointing out this spot of code, that was not my intent. It looks like I got away with that mistake in my testing because of another error in my code. If the version is anything besides 10.5, it will assign `perennial_no_canals` to a new shape file in a temporary directory, while assigning `stream_network_no_canals` to an in memory shape file. That meant that it wasn't actually overwriting in my testing, which explains why I saw it as working. I'll go ahead and change that and push.
